base/strings/annotated_io.jl

```diff
@@ -199,3 +199,78 @@ function _insert_annotations!(io::AnnotatedIOBuffer, annotations::Vector{RegionA
         push!(io.annotations, setindex(annotations[index], start+offset:stop+offset, :region))
     end
 end
+
+# NOTE: This is an interim solution to the invalidations caused
+# by the split styled display implementation. This should be
+# replaced by a more robust solution (such as a consolidation of
+# the type and method definitions) in the near future.
+module AnnotatedDisplay
+
+using ..Base: IO, SubString, AnnotatedString, AnnotatedChar, AnnotatedIOBuffer
+using ..Base: eachregion, invoke_in_world, tls_world_age
+
+# Write
+
+ansi_write(f::Function, io::IO, x::Any) = invoke(f, Tuple{typeof(io), supertype(typeof(x))}, io, x)
```

Comment:
This `invoke` is a little fishy - Can you justify the `supertype(typeof(x))`?

Reply:
Yup, this is because otherwise `print(::IO, ::AnnotatedString) -> ansi_write(print, ::IO, ::AnnotatedString) -> print(::IO, ::AnnotatedString) -> [...]`.

Reply:
This is a generic fallback though, so the supertype behavior should be appropriate for the general case, not just to resolve that specific recursion.

I'd prefer a stub `ansi_write(::Function, ::IO, ::AnnotatedString)` implementation that does what you want, since that's specifically where you know that this `invoke` is correct / intended

Reply:
It might be worth narrowing the `ansi_write` interface in general to just `x::Union{AnnotatedString,SubString{<:AnnotatedString},AnnotatedChar}`

Otherwise, I have a hard time defining what `ansi_write`'s behavior is for more general objects, since it seems to implicitly depend on a notion of "underlying string".

In general, it's strange that the stub implementation does not actually unwrap the underlying string like the full impl in StyledStrings - doing the `invoke` instead seems like a divergence from an interface that promises to pass the underlying string to `f` (i.e. it passes the original string to the underlying string's method of `f`, which is different)

Reply:
Do you think it would be better if we had this instead?

```julia
ansi_write(f::Function, io::IO, x::AnnotatedString) = f(io, x.string)
ansi_write(f::Function, io::IO, x::SubString{<:AnnotatedString}) = f(io, SubString(x.string.string, x.offset, x.ncodeunits, Val(:noshift)))
ansi_write(f::Function, io::IO, x::AnnotatedChar) = f(io, x.char)
```

Reply:
Yeah, that sounds good to me - seems like it lines up well with the implementation StyledStrings will provide later (just make sure that specificity doesn't prevent the override from working)

Reply:
Ah yes, that is the issue.

Reply:
Right, these "default" definitions are actually more specific than the implementation provided by `StyledStrings`: `ansi_write(f::F, io::IO, s::Union{<:AnnotatedString, SubString{<:AnnotatedString}})`

Hmmm, how bad is `ansi_write(f::Function, io::IO, x::Any) = f(io, String(x))`?

Reply:
That sounds fine to me, but you could also just adjust the Method definitions in StyledStrings to match the split

Reply:
I can't say I see how I could make it work with `ansi_write(f::Function, io::IO, x::AnnotatedString)`, since the method being called in `StyledStrings` has exactly that signature (actually, a slightly broader one, with a `x::Union{...}` ).

I'll go ahead and push `f(io, String(x))`.

Reply:
I was thinking you would split:
```julia
Base.AnnotatedDisplay.ansi_write(f::F, io::IO, s::Union{<:AnnotatedString, SubString{<:AnnotatedString}}) where {F <: Function} =
    _ansi_writer(f, io, s)
```

into:
```julia
Base.AnnotatedDisplay.ansi_write(f::F, io::IO, s::AnnotatedString) where {F <: Function} =  _ansi_writer(f, io, s)
Base.AnnotatedDisplay.ansi_write(f::F, io::IO, s::SubString{<:AnnotatedString}) where {F <: Function} = _ansi_writer(f, io, s)
```

but I forgot that these types of method replacements aren't permitted during pre-compilation

Anyway, the new fallback seems OK to me